### PR TITLE
[hue] Fix and improve error logging and status descriptions for API v2

### LINKED ISSUE #15511: [hue] Unhandled HTTP status code 207 (API v2)
I noticed this during system startup:
```
2023-08-27 17:18:58.698 [WARN ] [e.internal.handler.Clip2ThingHandler] - Command 'OFF' for thing 'hue:device:home:traadfri_white_bulb2', channel 'hue:device:home:traadfri_white_bulb2:brightness' failed with error 'Unexpected HTTP status '207''.
2023-08-27 17:18:58.762 [WARN ] [e.internal.handler.Clip2ThingHandler] - Command 'OFF' for thing 'hue:device:home:traadfri_color_bulb2', channel 'hue:device:home:traadfri_color_bulb2:color' failed with error 'Unexpected HTTP status '207''.
2023-08-27 17:18:58.795 [WARN ] [e.internal.handler.Clip2ThingHandler] - Command 'OFF' for thing 'hue:device:home:traadfri_white_bulb1', channel 'hue:device:home:traadfri_white_bulb1:brightness' failed with error 'Unexpected HTTP status '207''.
2023-08-27 17:18:58.845 [WARN ] [e.internal.handler.Clip2ThingHandler] - Command 'OFF' for thing 'hue:device:home:traadfri_color_bulb1', channel 'hue:device:home:traadfri_color_bulb1:color' failed with error 'Unexpected HTTP status '207''.
```

This was exposed by #15477 because status code is now checked and logged when different than 200.

## Expected Behavior
207 Multi-Status should be handled.

## Current Behavior
207 Multi-Status is not supported.

## Possible Solution
Documented [here](https://developers.meethue.com/develop/hue-api-v2/api-reference/#resource_light__id__put):

HTTP status code [207](http://httpstatus.es/207)
Body
**Media type:** application/json

**Type:** object

**Properties**

- **errors:** _required(array of Error)_
**Items:** Error
<pre>
description: required(string)
a human-readable explanation specific to this occurrence of the problem.
</pre>

- **data:** _required(array of ResourceIdentifierPut)_
**Items:** ResourceIdentifierPut
<pre>
rid: (string – pattern: ^[0-9a-f]{8}-([0-9a-f]{4}-){3}[0-9a-f]{12}$)
The unique id of the referenced resource

rtype: (one of device, bridge_home, room, zone, light, button, relative_rotary, temperature, light_level, motion, entertainment, grouped_light, device_power, zigbee_bridge_connectivity, zigbee_connectivity, zgp_connectivity, bridge, zigbee_device_discovery, homekit, matter, matter_fabric, scene, entertainment_configuration, public_image, auth_v1, behavior_script, behavior_instance, geofence, geofence_client, geolocation, smart_scene)
The type of the referenced resource
</pre>

## Steps to Reproduce
So far I have been able to reproduce this shortly after restarts. With trace logging enabled the first occurrence looks like this:
```
2023-08-27 18:02:45.178 [DEBUG] [e.internal.handler.Clip2ThingHandler] - 00000000-0000-0000-0000-000000000000 -> handleCommand() put resource id:11111111-1111-1111-1111-111111111111, type:light
2023-08-27 18:02:45.180 [DEBUG] [.internal.handler.Clip2BridgeHandler] - putResource() id:11111111-1111-1111-1111-111111111111, type:light
2023-08-27 18:02:45.182 [TRACE] [.hue.internal.connection.Clip2Bridge] - PUT https://192.168.0.251/clip/v2/resource/light/11111111-1111-1111-1111-111111111111 HTTP/2 >> {"type":"light","id":"11111111-1111-1111-1111-111111111111","on":{"on":false}}
2023-08-27 18:02:45.216 [TRACE] [.hue.internal.connection.Clip2Bridge] - HTTP/2 207 (Content-Type: application/json) << {"data":[{"rid":"11111111-1111-1111-1111-111111111111","rtype":"light"}],"errors":[{"description":"device (light) has communication issues, command (.on.on) may not have effect"}]}
2023-08-27 18:02:45.218 [DEBUG] [e.internal.handler.Clip2ThingHandler] - 00000000-0000-0000-0000-000000000000 -> handleCommand() error Unexpected HTTP status '207'
org.openhab.binding.hue.internal.exceptions.ApiException: Unexpected HTTP status '207'
        at org.openhab.binding.hue.internal.connection.Clip2Bridge.putResource(Clip2Bridge.java:1110) ~[?:?]
        at org.openhab.binding.hue.internal.handler.Clip2BridgeHandler.putResource(Clip2BridgeHandler.java:565) ~[?:?]
        at org.openhab.binding.hue.internal.handler.Clip2ThingHandler.handleCommand(Clip2ThingHandler.java:513) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
        at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:147) ~[?:?]
        at org.openhab.core.internal.common.InvocationHandlerSync.invoke(InvocationHandlerSync.java:59) ~[?:?]
        at jdk.proxy1151.$Proxy1279.handleCommand(Unknown Source) ~[?:?]
        at org.openhab.core.thing.internal.profiles.ProfileCallbackImpl.handleCommand(ProfileCallbackImpl.java:85) ~[?:?]
        at org.openhab.core.thing.internal.profiles.SystemDefaultProfile.onCommandFromItem(SystemDefaultProfile.java:48) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
        at org.openhab.core.internal.common.AbstractInvocationHandler.invokeDirect(AbstractInvocationHandler.java:147) ~[?:?]
        at org.openhab.core.internal.common.Invocation.call(Invocation.java:52) ~[?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
        at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

Pretty-printed payload:
```json
{
	"data": [
		{
			"rid": "11111111-1111-1111-1111-111111111111",
			"rtype": "light"
		}
	],
	"errors": [
		{
			"description": "device (light) has communication issues, command (.on.on) may not have effect"
		}
	]
}
```

## Your Environment
* Version used: 4.0.2 with binding compiled after merging #15477

### PR DESCRIPTION
**Pull request by jlaur**
When sending a command to an unreachable bulb this **warning** was logged:
```
[WARN] Command 'OFF' for thing 'hue:device:home:innr3', channel 'hue:device:home:innr3:switch' failed with error 'Unexpected HTTP status '207''.
```

This has been changed to log level **information** with the errors provided in the response:
```
[INFO] Command 'OFF' for thing 'hue:device:home:innr3', channel 'hue:device:home:innr3:switch' succeeded with errors: device (light) has communication issues, command (.on.on) may not have effect
```

Or this, in case of multiple errors:
```
[INFO] Command 'ON' for thing 'hue:device:home:traadfri_color_bulb2', channel 'hue:device:home:traadfri_color_bulb2:color' succeeded with errors: device (light) has communication issues, command (.on.on) may not have effect; device (light) has communication issues, command (.dimming.brightness) may not have effect; device (light) has communication issues, command (.color.xy) may not have effect
```

Prior to #15477 nothing was logged with default log level (log level **debug** required).

The reason why these errors are not strictly treated as errors is that the PUT request actually succeeds, at least partially (hence "may not have effect"). For example, for `dimming-only`, the requested level is immediately visible in the Hue app even when the bulb is off. For some supported bulb types the new values might be sent to the bulb when it will come online again.

Fixes #15511
Regression of #15477

Also improves generic error messages to obtain more information in case of issues - see https://github.com/openhab/openhab-addons/issues/15460#issuecomment-1699839595

### CODE REVIEW

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resources.java
Comment (andrewfg) [inline bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resources.java:31]: Could we please keep the old Collectors.toList() syntax here? I am still maintaining a back port to OH v3.x..

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java
Comment (andrewfg) [inline bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java:1110]: As I don’t have access to a PC can you please confirm that isSuccess() returns true for all 20x errors?

**Review by andrewfg — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java
Comment (jlaur) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java:1110]: It does exactly that, here is the implementation:
```java
    public static boolean isSuccess(int code)
    {
        return ((200 <= code) && (code <= 299));
    }
```

Comment (andrewfg) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/connection/Clip2Bridge.java:1110]: Ok. Thanks.

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resources.java
Comment (jlaur) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resources.java:31]: This is not just syntax. See https://rules.sonarsource.com/java/RSPEC-6204/

In general I consider this topic already covered by https://github.com/openhab/openhab-addons/pull/15419#discussion_r1292739917 and https://github.com/openhab/openhab-addons/pull/15390#discussion_r1292741086.

Feel free to continue backporting, of course. I don't know how you approach it from a practical perspective, but it might make your life easier if you consider `main` as upstream to your 3.x branch, i.e. cherry-pick commits into it and then apply Java and openHAB version compatibility patches after that when needed.

Comment (andrewfg) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resources.java:31]: The prior code worked fine. There is no reason to change it.

Comment (andrewfg) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resources.java:31]: PS I know that I have no influence in this discussion since you have always the power to overrule. But if you do insist on overruling then I think you should at least be self consistent and change ALL the cases in my code where I have used Collectors.toList .. there is no point in being half pure.

Comment (jlaur) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Resources.java]: I was applying the "boy scout rule" here. However, I have now reverted it from this PR since this refactoring was only related by "other changes close by". This can definitely be fixed more consistently in a follow-up PR and you can simply omit that resulting commit from your backporting efforts.

**Review by andrewfg — APPROVED** [review]
LGTM thankyou.

**Comment by andrewfg** [comment]
@jlaur your last commit goes far beyond the original title of this PR, and potentially impacts some quite fundamental functionality, so before I review it, perhaps you can say why you changed it in the way that you did?

**Comment by jlaur** [comment]
> perhaps you can say why you changed it in the way that you did?

To have better error information available in the status description, which can help pinpoint (rare) issues more easily. This would otherwise require potentially long-running reproduction attempts with **trace** logging enabled (which is very verbose).

As an example, I observed an error message like this: _"An unexpected exception 'Error sending request' occurred."_

This is very generic and without much useful information. All thrown `ApiException`s are like this, but most of them actually also provides the original exception which could give more insight into the root cause:

image

The status update for **COMMUNICATION_ERROR** was extracted to `setStatusOfflineWithCommunicationError` in order to reuse the new logic for providing the inner exception message also.

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java
Comment (andrewfg) [inline bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:178]: Surely retryApplicationKey should be false here as the registration will have succeeded. Or?

**Review by andrewfg — CHANGES_REQUESTED** [review]

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java
Comment (jlaur) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:178]: You might be right, but I didn't want to change behavior in context of this PR. If we can get this PR merged (to avoid conflicts), would you do a follow-up PR for this?

Comment (andrewfg) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:178]: As I said in my prior post, this is quite critical functionality, so we should not merge hastily until it has been fully tested. Which I cannot do until I return next week. I was happy with the PR until your last commit, so if you want to merge quickly then I suggest to revert that last commit.

Comment (jlaur) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:178]: Let's not rush, it can wait.

Comment (jlaur) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:178]: Resolving comment since fixing application key logic is outside the scope of this PR.

Comment (andrewfg) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:178]: Ok. I checked it. The use of `retryApplicationKey` is correct, so no changes required.

**Review by andrewfg — APPROVED** [review]
LGTM

**Comment by jlaur** [comment]
@openhab/add-ons-maintainers - after merging this, please cherry-pick the commit into the `4.0.x` branch since it fixes a regression in the upcoming hotfix.

**Comment by jlaur** [comment]
@lolodomo - can you review this?

**Comment by kaikreuzer** [comment]
> Prior to https://github.com/openhab/openhab-addons/pull/15477 nothing was logged with default log level (log level debug required).

So why didn't you do it this way again, but decided now for INFO level?
After all, we never log any info when commands are sent to things that are OFFLINE. And as you explain, in this case here it might actually do something (sending the value to the bridge, while not to the bulb), so everything actually works as expected.

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by lolodomo** [comment]
> > Prior to #15477 nothing was logged with default log level (log level debug required).
> 
> So why didn't you do it this way again, but decided now for INFO level? After all, we never log any info when commands are sent to things that are OFFLINE. And as you explain, in this case here it might actually do something (sending the value to the bridge, while not to the bulb), so everything actually works as expected.

I am very sorry, Kai, I did not see your comment (we were in parallel on this PR).

**Comment by kaikreuzer** [comment]
No worries, in case we decide to change the logging level that would just be a one line change.

**Comment by kaikreuzer** [comment]
@lolodomo Will you take care of cherry-picking?

**Comment by lolodomo** [comment]
> @lolodomo Will you take care of cherry-picking?

Please do it, I am not yet familiar at all with that.

**Comment by kaikreuzer** [comment]
Ok.

**Comment by lolodomo** [comment]
@jlaur : even if I merged your PR a little too early, please continue the discussion with Kai ;)

**Comment by kaikreuzer** [comment]
Cherry-picked.

**Comment by jlaur** [comment]
> So why didn't you do it this way again, but decided now for INFO level?\nAfter all, we never log any info when commands are sent to things that are OFFLINE. And as you explain, in this case here it might actually do something (sending the value to the bridge, while not to the bulb), so everything actually works as expected.

I tend to agree. I was divided because it also applies to online things where the command only partially succeesed. @andrewfg - any thoughts? I would be okay wirh debug level as well.

**Comment by andrewfg** [comment]
> it also applies to online things where the command only partially succeeded. @andrewfg - any thoughts?

The specific use case is that the Hue Bridge is online (so it can definitely receive commands) yet the device itself has communication issues (so the bridge cannot guarantee that the device immediately received the respective command). It is definitely not ERROR or WARN; and either INFO or DEBUG would make sense. It would not normally happen very frequently.

**Comment by andrewfg** [comment]
@jlaur for info, can you please let me know when 4.0.3 is expected to release?

**Comment by lolodomo** [comment]
Today, build is already started (and maybe terminated).

**Comment by jlaur** [comment]
@andrewfg - https://ci.openhab.org/view/Release%20Jobs/job/openhab-patch-release/36/

**Pull request merged**
